### PR TITLE
Unify code path for metaseq and metaseq-internal

### PR DESCRIPTION
**Pull request by Xirider**
Currently we have 2 versions of sweep.py and slurm.py: One version is used when a opt_baseline script is run from metaseq and the other version is used with sweep_baseline from metaseq-internal. Maintaining both versions adds unnecessary complexity to the code base and makes testing more difficult.

This PR brings most features of the sweep and slurm file from metaseq-internal to metaseq, in preparation of deleting these files in metaseq-internal.

Here some notes:
- brought in the tombstone feature
- metaseq-internal had a wrapper (to log some worker info) around the training script "train_wrapper.py" that I now moved to the train.py file
- there was some shuffling around in the path logic for the train command in slurm.py, so that it will now work independently of the user's working directory

Some things were not brought in (i.e. flags without use).
For a number of other features that I included I'm not sure if they are actually used currently:
- post_cmds
- container_image and container_save
- array_length
- args.dep and args.sequential
Should these stay?

Issue: https://github.com/facebookresearch/metaseq/issues/472
Internal PR: https://github.com/fairinternal/metaseq-internal/pull/558

Testing: 
`python metaseq/launcher/opt_baselines.py --prefix train.8m --model-size 8m --checkpoints-dir ./test-checkpoint --tensorboard-logdir ./test-checkpoint --num-trials 1 --azure --num-gpus 4 --num-nodes 1 --seed 1 --circleci --local --disable-validation --max-epoch 100 --max-update 100`

### CODE REVIEW

File: metaseq/launcher/slurm.py [hunk 1]
```diff
@@ -461,11 +580,19 @@ def launch_train(args, grid, grid_product, dry_run, postprocess_hyperparams):
...
             else:
                 write_git_commit(train_log)
+                if args.rsc:
+                    klist(train_log)
                 with open(train_log, "a") as train_log_h:
                     job_id, stdout = run_batch(env, sbatch_cmd_str, sbatch_cmd)
                     print(stdout, file=train_log_h)
         if job_id is not None:
             print("Launched {}".format(job_id))
+        if args.sequential and not args.local and job_id is not None:
```
Comment (stephenroller) [inline metaseq/launcher/slurm.py:590]: what is this lol

File: metaseq/launcher/sweep.py [hunk 2]
```diff
@@ -227,15 +252,24 @@ def _get_args(add_extra_options_func=None, input_args: Optional[List[str]] = Non
...
         help="save tensorboard logs in <tensorboard-logdir>/<prefix>.<save_dir_key>",
     )
+    parser.add_argument(
+        "-ts",
+        "--tombstonable",
+        type=bool,
+        default=False,
+        help="""make the job killable by writing a
+                 tombstone 'tombstone_<job_id>' file to user's home ditectory
+                (/shared/home/$USER)""",
```
Comment (stephenroller) [inline metaseq/launcher/sweep.py:262]:
```suggestion
        help=(
            "make the job killable by writing a "
            "tombstone 'tombstone_<job_id>' file to user's home directory "
            "(/shared/home/$USER)"
        ),
```

File: metaseq/launcher/tombyard.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+import datetime
+import os
+import time
+
+
+def tombstones_procedure(
+    job_id,
+    dirstones,
+    period_before_tombstone_detected=datetime.timedelta(seconds=60),
+    period_after_tombstone_detected=datetime.timedelta(seconds=3),
```
Comment (stephenroller) [inline metaseq/launcher/tombyard.py:16]: let's move the defaults into constants at the top

File: metaseq/launcher/tombyard.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        if tombstone_detected:
+            print(f".. scanceling the job and its current squeue.state is {status}")
+        if status not in job_state_codes:
+            print(f"Done scanceling the job. Its squeue.state now is: {status}")
+            return
+        if not tombstone_detected:
+            for tombstone_name in dirstones["scancel"]:
+                if os.path.exists(tombstone_name):
+                    print(
+                        f"""tombstones_procedure has detected file {tombstone_name}.
```
Comment (stephenroller) [inline metaseq/launcher/tombyard.py:61]: fix awkward whitespace similar to other suggestion, here and others

File: metaseq/launcher/tombyard.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+                          scontrol requeuehold {job_id} will be called once.
+                         remove the file {tombstone_name} within the next {period_before_tombstone_detected}
+                         for it not to trigger the same command again"""
+                    )
+                    _ = os.popen(
+                        f"""
+                        scontrol requeuehold {job_id}
+                         """
+                    ).read()
```
Comment (stephenroller) [inline metaseq/launcher/tombyard.py:79]:
```suggestion
                    _ = os.popen(f"scontrol requeuehold {job_id}").read()
```

File: metaseq/launcher/tombyard.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+                    scancel {job_id}
+                    """
+            ).read()
+        time.sleep(period.total_seconds())
+
+
+def tombstones(job_id, period=datetime.timedelta(seconds=60), dirstones=None):
+    if dirstones is None:
+        dirstones = {"scancel": [], "requeuehold": [], "release": []}
+        azure_base = "/shared/home"
```
Comment (stephenroller) [inline metaseq/launcher/tombyard.py:105]: we're hardcoding here i believe

File: metaseq/launcher/tombyard.py [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+            dirstones["release"].append(
+                os.path.join(azure_base, userdir, f"release_{job_id}")
+            )
+
+    for directive in ["scancel", "requeuehold", "release"]:
+        if directive not in dirstones.keys():
+            dirstones[directive] = []
+
+    # start a process that monitors
+    ctx = mp.get_context("spawn")
```
Comment (stephenroller) [inline metaseq/launcher/tombyard.py:122]: this reads like the launch node is where this monitoring happens, and I'm not sure that's what we want. I think you just ported this code so let's make a task to fix this

File: metaseq/launcher/slurm.py [hunk 8]
```diff
@@ -251,18 +279,45 @@ def gen_srun_command_and_str(args, save_dir_key, train_log, train_stderr, train_
         ]
         base_srun_cmd += ["-x", excluded_hosts] if excluded_hosts is not None else []
         base_srun_cmd += ["-w", included_hosts] if included_hosts is not None else []
+    if args.container_image is not None:
```
[the remark this answers is not in the document]
Comment (Xirider) [reply metaseq/launcher/slurm.py:282]: 🤷  I guess I'll just remove it.

File: metaseq/launcher/slurm.py (hunk 1 above)
Comment (Xirider) [reply metaseq/launcher/slurm.py:590]: Do you know if the other things here are actually still used by people?
post_cmds
container_image and container_save - will remove
array_length
args.dep and args.sequential - will remove

Comment (stephenroller) [reply metaseq/launcher/slurm.py:590]: Containers might have been an external contributor except then why is it in metaseq-internal? Can we gut blame and try to figure out the why here? I'm afraid I may be missing context

Comment (stephenroller) [reply metaseq/launcher/slurm.py:590]: Containers might have been an external contributor except then why is it in metaseq-internal? Can we gut blame and try to figure out the why here? I'm afraid I may be missing context

File: metaseq/launcher/tombyard.py (hunk 7 above)
Comment (Xirider) [reply metaseq/launcher/tombyard.py:122]: Yes, I just moved the file to allow for a unified way of importing, so I did not look at the details. But I will fix the whitespace and create an separate issue for this.

File: metaseq/launcher/slurm.py (hunk 1 above)
Comment (Xirider) [reply metaseq/launcher/slurm.py:590]: https://github.com/fairinternal/metaseq-internal/commit/2fb31531c85d0f82eaaea9d4d463e60afc4a74c5
Seems like this was for an old training run on azure. So I'll remove it as well.

File: metaseq/launcher/slurm.py [hunk 9]
```diff
@@ -366,6 +427,31 @@ def write_git_commit(train_log):
         print(git_commit.rstrip(), file=train_log_h)
 
 
+def klist(train_log):
```
[the remark this answers is not in the document]
Comment (Xirider) [reply metaseq/launcher/slurm.py]: Moved this out now into a new file for internal functions for now.

File: metaseq/launcher/tombyard.py (hunk 6 above)
Comment (Xirider) [reply metaseq/launcher/tombyard.py]: changed it to import the correct base path from constants.py

**Comment by Xirider** [comment]
Fixed the hardcoded base directory in tombstones feature for rsc.

Can someone approve this here and the internal PR as well? Thanks!

File: metaseq/launcher/slurm.py [hunk 10]
```diff
@@ -288,61 +331,63 @@ def gen_sbatch_command_and_str(
...
-        comment += (
-            f", OSS Code Location: {oss_destination}"
-            if comment
-            else f"OSS Code Location: {oss_destination}"
-        )
+        if args.snapshot_code:
+            comment += f", OSS Code Location: {oss_destination} Internal Code Location: {internal_destination}"
         sbatch_cmd += ["--comment", comment]
-
+    elif args.snapshot_code:
```
Comment (stephenroller) [inline metaseq/launcher/slurm.py:351]: Nit: This is redundant with above. As annoying as it is, I preferred the `comment = ""` bit as it was

File: metaseq/launcher/slurm.py [hunk 11]
```diff
@@ -461,11 +539,16 @@ def launch_train(args, grid, grid_product, dry_run, postprocess_hyperparams):
...
+                if args.rsc:
+                    internal.klist(train_log)
                 with open(train_log, "a") as train_log_h:
                     job_id, stdout = run_batch(env, sbatch_cmd_str, sbatch_cmd)
                     print(stdout, file=train_log_h)
         if job_id is not None:
             print("Launched {}".format(job_id))
+        if hasattr(args, "tombstonable"):
+            if args.tombstonable:
+                tombstones(job_id=job_id, base_dir=args.default_prefix)
```
Comment (stephenroller) [inline metaseq/launcher/slurm.py:551]: is default_prefix the right thing here? Should it be the actual prefix of the job?

I'm still not sure we're checking for tombstone logic in the right place. I'm not sure this will work on RSC as is

File: metaseq/launcher/tombyard.py [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+    "SIGNALING",
+    "SPECIAL_EXIT",
+    "STAGE_OUT",
+    "STOPPED",
+    "SUSPENDED",
+    "TIMEOUT",
+]
+
+
+def tombstones_procedure(
```
Comment (stephenroller) [inline metaseq/launcher/tombyard.py:39]: discussed offline, we'll make tombstones a separate PR

File: metaseq/launcher/slurm.py (hunk 11 above)
Comment (Xirider) [reply metaseq/launcher/slurm.py:551]: The name is just for consistency, but it is the actual prefix of the job. It results in the right path that Susan wrote here: https://github.com/fairinternal/metaseq-internal/pull/564#issuecomment-1300799020

**Review by stephenroller — APPROVED** [review]
discussed getting rid of post commands and then let's go

File: metaseq/launcher/slurm.py (hunk 11 above)
Comment (Xirider) [reply metaseq/launcher/slurm.py]: Changed naming of the arg to make it clearer what it is.

**Comment by Xirider** [comment]
Thanks, Stephen!
